### PR TITLE
Period load/drop/broadcast rules should include the future by default

### LINKED ISSUE #5869: Period load rules should include the future by default
We have seen problem scenarios like the following:

1. A user has load rules like: load 30 days, drop forever.
2. The user loads some data from slightly in the future (maybe some clocks are running a bit fast or slow) using streaming ingestion. This creates a segment with an interval that is in the future.
3. The coordinator disables the segment immediately upon noticing it (since it is not within the last 30 days).
4. The Kafka tasks time out during handoff (because the segments are never loaded).
5. And after that timeout, the data that was slightly in the future is _still_ not available!

It would be better, I think, if period-based load rules included an unbounded time into the future too. I bet this lines up better with users' intuitive idea of what "retain 30 days" means. Of course it could be changeable with some optional parameters (or a different kind of rule?) but I think the default behavior should be the intuitive one.

NB: A workaround to this is to use rules like "load 30 days" followed by "drop 1000 years" followed by "load forever". That has the same effect as what I am suggesting for the new default behavior: we'll drop anything older than 30 days, but retain the last 30 days and also anything that is in the future.

Related, but different: #5868.

### PR DESCRIPTION
**Pull request by QiuMM**
Fix #5869 .

### CODE REVIEW

**Comment by QiuMM** [comment]
#5869

File: server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java [hunk 1]
```diff
@@ -94,7 +104,9 @@ public boolean equals(Object o)
     if (!Objects.equals(period, that.period)) {
       return false;
     }
-
+    if (!Objects.equals(includeFuture, that.includeFuture)) {
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java]: `includeFuture` can't be null. `if (includeFuture != that.includeFuture)`?

File: server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java [hunk 2]
```diff
@@ -108,29 +111,61 @@ public void testAppliesToPartialOverlap()
...
+    Assert.assertFalse(
+        notIncludeFutureRule.appliesTo(
+            builder.interval(new Interval(now.plusDays(1), now.plusDays(2))).build(),
+            now
+        )
+    );
+  }
+
+  @Test
+  public void testSerdeNull() throws Exception
```
Comment (jihoonson) [inline server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java]: Why did you change the test name?

File: server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java [hunk 3]
```diff
@@ -108,29 +111,61 @@ public void testAppliesToPartialOverlap()
...
     Assert.assertEquals(rule.getPeriod(), ((PeriodLoadRule) reread).getPeriod());
+    Assert.assertEquals(rule.isIncludeFuture(), ((PeriodLoadRule) reread).isIncludeFuture());
+    Assert.assertEquals(PeriodLoadRule.DEFAULT_INCLUDE_FUTURE, rule.isIncludeFuture());
     Assert.assertEquals(rule.getTieredReplicants(), ((PeriodLoadRule) reread).getTieredReplicants());
     Assert.assertEquals(ImmutableMap.of(DruidServer.DEFAULT_TIER, DruidServer.DEFAULT_NUM_REPLICANTS), rule.getTieredReplicants());
   }
 
   @Test
-  public void testMappingNullTieredReplicants() throws Exception
+  public void testMappingNull() throws Exception
```
Comment (jihoonson) [inline server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java]: Same. Why did you change the name?

File: docs/content/operations/rule-configuration.md [hunk 4]
```diff
@@ -73,9 +74,10 @@ Period load rules are of the form:
 
 * `type` - this should always be "loadByPeriod"
 * `period` - A JSON Object representing ISO-8601 Periods
+* `includeFuture` - A JSON Boolean indicating whether the load period should include the future. Default is true.
```
[the remark this answers is not in the document]
Comment (QiuMM) [reply docs/content/operations/rule-configuration.md]: like `A JSON Boolean indicating whether the load period should include the future. This property is optional, Default is true.` ?

File: server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java (hunk 1 above)
Comment (QiuMM) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java]: Good, I'll change it.

File: server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java (hunk 2 above)
Comment (QiuMM) [reply server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java]: Because I have also test `Serde` null of `includeFuture` in this test case.

File: server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java (hunk 3 above)
Comment (QiuMM) [reply server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java]: Same reason.

**Comment by QiuMM** [comment]
I'll add more commits to adjust coordinator console after #6415 being merged, because there may be some code conflicts between these two pr.

File: docs/content/operations/rule-configuration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply docs/content/operations/rule-configuration.md]: Sounds good.

File: server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java (hunk 2 above)
Comment (jihoonson) [reply server/src/test/java/org/apache/druid/server/coordinator/rules/PeriodLoadRuleTest.java]: Ok. Then, please add some comments about what null variables are being tested in this test. Same for the below test.

**Comment by QiuMM** [comment]
@jihoonson have updated, you can go on reviewing this PR. Below is the UI when configure the period rules:
lark20181026-125716

File: docs/content/operations/rule-configuration.md [hunk 5]
```diff
@@ -73,9 +74,10 @@ Period load rules are of the form:
 
 * `type` - this should always be "loadByPeriod"
 * `period` - A JSON Object representing ISO-8601 Periods
+* `includeFuture` - A JSON Boolean indicating whether the load period should include the future. This property is optional, Default is true.
 * `tieredReplicants` - A JSON Object where the keys are the tier names and values are the number of replicas for that tier.
 
-The interval of a segment will be compared against the specified period. The rule matches if the period overlaps the interval.
+The interval of a segment will be compared against the specified period. The period is from some time in the past to the future or to the current time, which depends on `includeFuture` is true or false. The rule matches if the period `overlaps` the interval.
```
Comment (jihoonson) [inline docs/content/operations/rule-configuration.md]: The backtick (`) is for code blocks. How about _italics_(single asterisk or underscore) or __bold__ (double asterisk or underscore)?

File: docs/content/operations/rule-configuration.md [hunk 6]
```diff
@@ -120,14 +122,16 @@ Period drop rules are of the form:
...
+  "includeFuture" : true
 }
 ```
 
 * `type` - this should always be "dropByPeriod"
 * `period` - A JSON Object representing ISO-8601 Periods
+* `includeFuture` - A JSON Boolean indicating whether the load period should include the future. This property is optional, Default is true.
 
-The interval of a segment will be compared against the specified period. The period is from some time in the past to the current time. The rule matches if the period contains the interval.
+The interval of a segment will be compared against the specified period. The period is from some time in the past to the future or to the current time, which depends on `includeFuture` is true or false. The rule matches if the period `contains` the interval. This drop rule always dropping recent data.
```
Comment (jihoonson) [inline docs/content/operations/rule-configuration.md]: Same. How about _italics_(single asterisk or underscore) or **bold** (double asterisk or underscore)?

File: docs/content/operations/rule-configuration.md [hunk 7]
```diff
@@ -173,15 +177,17 @@ Period broadcast rules are of the form:
...
 }
 ```
 
 * `type` - this should always be "broadcastByPeriod"
 * `colocatedDataSources` - A JSON List containing data source names to be co-located. `null` and empty list means broadcasting to every node in the cluster.
 * `period` - A JSON Object representing ISO-8601 Periods
+* `includeFuture` - A JSON Boolean indicating whether the load period should include the future. This property is optional, Default is true.
 
-The interval of a segment will be compared against the specified period. The period is from some time in the past to the current time. The rule matches if the period contains the interval.
+The interval of a segment will be compared against the specified period. The period is from some time in the past to the future or to the current time, which depends on `includeFuture` is true or false. The rule matches if the period `overlaps` the interval.
```
Comment (jihoonson) [inline docs/content/operations/rule-configuration.md]: Same. How about _italics_(single asterisk or underscore) or **bold** (double asterisk or underscore)?

File: server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java [hunk 8]
```diff
@@ -32,17 +32,21 @@
 public class PeriodBroadcastDistributionRule extends BroadcastDistributionRule
 {
   static final String TYPE = "broadcastByPeriod";
+  public static final boolean DEFAULT_INCLUDE_FUTURE = true;
```
Comment (jihoonson) [inline server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java]: Can be private.

File: server/src/main/resources/static/old-console/js/rules-0.0.2.js [hunk 9]
```diff
@@ -114,6 +114,7 @@ function makeLoadByInterval(rule) {
 function makeLoadByPeriod(rule) {
   var retVal = "";
   retVal += "<span class='rule_label'>period</span><input type='text' name='period' " + "value='" + rule.period + "'/>";
+  retVal += "<span class='rule_label'>includeFuture</span><input type='text' name='includeFuture' " + "value='" + true + "'/>";
```
Comment (jihoonson) [inline server/src/main/resources/static/old-console/js/rules-0.0.2.js]: This looks a legacy UI. Do you have a plan to update the recent UI as well?

Comment (QiuMM) [reply server/src/main/resources/static/old-console/js/rules-0.0.2.js]: Yes, I'll open a PR in [druid-console](https://github.com/druid-io/druid-console) to update the UI.

File: server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java (hunk 8 above)
Comment (jihoonson) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java]: Oh, it looks that I was wrong. It's being used in `PeriodLoadRuleTest`. Maybe package-private?

Comment (jihoonson) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java]: Ah there're several `DEFAULT_INCLUDE_FUTURE`. Please fix them properly.

Comment (QiuMM) [reply server/src/main/java/org/apache/druid/server/coordinator/rules/PeriodBroadcastDistributionRule.java]: Sorry, It's my fault. I have never checked before push.
